### ISSUE TITLE
Run the after.insert hook if an upsert turns out to be an insert.

### ISSUE THREAD

**Issue by jamesgibson14** [matb33/meteor-collection-hooks#16]

**Comment by DenisGorbachev** [comment]
:+1: for after.insert, just ran into this issue.

**Comment by tagrudev** [comment]
any development on this one ?

**Comment by matb33** [comment]
Yes -- I had started and almost finished... but was pulled away for other work. I ended up just committing and pushing a wip under the `upsert` branch. I'll get back to it eventually

**Comment by matb33** [comment]
I've got a beta version of the upsert hooks implemented in the `upsert` branch. I just need to write tests for the sync version of upsert. That being said, anyone interested in trying their hand at writing these? I'm not sure when I'll get back to this!

**Comment by mizzao** [comment]
@matb33 I'll write them, can you summarize exactly what should be tested? I also just realized I've been using upserts on a hooked collection for a while now and it has mostly been working as expected.  Is it just the update hook that is being called?

**Comment by rclai** [comment]
Yes, it only triggers update hooks.

**Comment by mizzao** [comment]
In the branch that @matb33 is working on, it seems that `before.update` will no longer be triggered at all, while `after.insert` and `after.update` will be triggered appropriately based on the outcome. (b69bfdacba68877311dc6dde0246651522e39f68). This may be a breaking change that we should warn about (e.g. it is for https://github.com/mizzao/meteor-partitioner).

So what is supposed to happen if someone calls `update` with `{upsert: true}`? Is that going to fire a `before.update` or `before.upsert` hook?

**Comment by matb33** [comment]
Currently, if someone calls `update` with `{upsert: true}` that will call `before.update`. If someone calls `upsert` that will call `before.upsert`.

I'm on the fence at this point as to how it should behave. I'm tempted to treat `update` with `{upsert: true}` as a special case and have it call `before.upsert`. Otherwise, we end up with the problem that calling `update` with `{upsert: true}` and an `insert` occurs, then no `after.XXX` hook will trigger.

**Comment by matb33** [comment]
I pushed a change to the `upsert` branch that has a call to `update` with `{upsert:true}` use the upsert advice.

**Comment by rclai** [comment]
So if we do an upsert that will insert, it will do:

`before.upsert` -> `after.insert`

And if we do an upsert that will update it will do:

`before.upsert` -> `after.update`

Is this right?

**Comment by matb33** [comment]
I didn't bother with an `after.upsert`, although it sounds like it would make sense...

By the way, it seems that treating `update` with `{upsert: true}` the same as an `upsert` won't work, as when one calls it that way, Meteor makes a distinction and does NOT return `insertedId`. That means we can't tell if it ended up doing an `insert` or an `update.

As per the Meteor docs:
```
upsert is the same as calling update with the upsert option set to true, except that the return value of upsert is an object that contain the keys numberAffected and insertedId. (update returns only the number of affected documents.)
```

**Comment by matb33** [comment]
@mizzao so after having worked on the upsert branch today, the tests that need to be written are the sync version of the upsert tests, and writing documentation in the README

Let me know if you can handle some or all of it, anything is appreciated! I'm actively trying to lower the bus factor on this project ;)

**Comment by mizzao** [comment]
Yep, happy to take a look at this when I have a moment.

**Comment by matb33** [comment]
I ended up writing some tests for sync upsert (which uncovered a bug), and added a bit of documentation to the README. I've published the package as 0.8.0 (upped the minor revision number to reflect minor breaking change with before.update vs before.upsert)
